src/rules/color-hex-notation/index.js

```diff
@@ -0,0 +1,38 @@
+import {
+  report,
+  ruleMessages,
+  styleSearch
+} from "../../utils"
+
+export const ruleName = "color-hex-notation"
+
+export const messages = ruleMessages(ruleName, {
+  expected: (c, h) => `Expected ${c} hex color ${h}`,
+})
+
+/**
+ * @param {"lowercase"|"uppercase"} expectation
+ */
+export default function (expectation) {
+  return (root, result) => {
+    root.eachDecl(decl => {
+      const value = decl.value
+
+      styleSearch({ source: value, target: "#" }, match => {
+
+        const hexValue = /^#[0-9A-Za-z]+/.exec(value.substr(match.startIndex))[0]
+
+        if (expectation === "lowercase" && (/^#(?:[0-9a-z]+)$/.test(hexValue))) { return }
+
+        if (expectation === "uppercase" && (/^#(?:[0-9A-Z]+)$/.test(hexValue))) { return }
+
```

Comment:
I wonder if there might be a way to avoid doing two regexp checks. What about something like:

```js
if (expectation === "lowercase" && hexValue === hexValue.toLowerCase()) { return }
```

Does that work?